### PR TITLE
Add labels in a more principled way.

### PR DESCRIPTION
**Pull request by csilvers**
## Summary:
Before, I was just munging the label to the name, so I wouldn't have
to change a lot of code.  This didn't work because the name was used
in places where the code needs it to actually be a name (e.g. when
telling github what reviewers to add to the PR).

So now I store names and labels separately, and just print the label
explicitly when emitting the comment text.  This also gives me more
control over where the label goes; all in all a more maintainable
arrangement.

Issue: https://khanacademy.atlassian.net/browse/FEI-5970

## Test plan:
yarn jest

### CODE REVIEW

File: src/runOnPullRequest.js [hunk 1]
```diff
@@ -22,6 +22,8 @@ import {
     MATCH_COMMENT_HEADER_REGEX,
 } from './constants';
 
+type NameToLabelToFiles = {[name: string]: {[label: string]: string[], ...}, ...};
```
Comment (lillialexis) [inline src/runOnPullRequest.js:25]: Before, if there were multiple matching rules, I think it would do:
```
{
  lilli: [fileFromR1, fileFromR2, fileFromR2]
...
}
```

Then in the comment it would do:
```
lilli because fileFromR1, fileFromR2, fileFromR2
```

But now if there are multiple matching rules, I think I would want:
```
lilli (rule1): fileFromR1
lilli (rule2): fileFromR2, fileFromR2
```

I'm not clear if this structure gets us to that.

Comment (csilvers) [reply src/runOnPullRequest.js:25]: Yes, it does!  You can see in some of the test output.

Comment (lillialexis) [reply src/runOnPullRequest.js:25]: Maybe it does, but the syntax is starting to twist my head...

File: src/runOnPush.js [hunk 2]
```diff
@@ -5,16 +5,19 @@ import {execCmd} from './execCmd';
 import {ownerAndRepo, extraPermGithub, type Context} from './setup';
 import {PUSH, GERALD_COMMIT_COMMENT_HEADER} from './constants';
 
-const makeCommitComment = async (
-    peopleToFiles: {[string]: Array<string>, ...},
-    commitSHA: string,
-) => {
-    const names: string[] = Object.keys(peopleToFiles);
-    if (peopleToFiles && names.length) {
+type NameToLabelToFiles = {[name: string]: {[label: string]: string[], ...}, ...};
```
Comment (lillialexis) [inline src/runOnPush.js:8]: Can we put this somewhere shared so as to not dupe?

File: src/runOnPullRequest.js (hunk 1 above)
Comment (csilvers) [reply src/runOnPullRequest.js:25]: If js allowed non-string map keys, I could do a map from `(person, label)` to `[files]`, but it doesn't, so this is the best I could do...

File: src/runOnPush.js (hunk 2 above)
Comment (csilvers) [reply src/runOnPush.js:8]: If you want.  I like having it in each place so it's easier to find when reading the file; flow won't let them get out of sync.  But I know some people feel very militant about DRY!

File: src/utils.js [hunk 3]
```diff
@@ -36,42 +36,47 @@ import {
...
+    peopleToLabelToFiles: NameToLabelToFiles,
     header: CommentHeaders,
     tagPerson: boolean,
 }) => {
-    const names: string[] = Object.keys(peopleToFiles);
-    if (names.length) {
-        let body = `<details>\n<summary><b>${header}</b></summary>\n\n`;
-        names.forEach((person: string) => {
-            const files = peopleToFiles[person];
+    const names: string[] = Object.keys(peopleToLabelToFiles);
```
Comment (lillialexis) [inline src/utils.js:58]: I bet this impacts the #remove-me functionality. Have you checked that? This is big enough that it's fine to do that part in another PR if it's easier.

**Review by lillialexis — APPROVED** [review]
Tentatively approving but we will want to check the #remove-me functionality.

You can also just test things in the Gerald repo, I think, if you want to avoid the long chain of gerald → actions → webapp.

There's also some local tester utility, but I have no idea what it does or how it works. I just saw it when picking though the code.

**Comment by lillialexis** [comment]
Here's Gerald's comment: https://github.com/Khan/gerald/pull/90#issuecomment-2471741967 for more context on self-testing and #remove-me

File: src/utils.js (hunk 3 above)
Comment (csilvers) [reply src/utils.js:59]: I did check that, at least by reading the code.  My changes do not affect `#remove-me` as far as I can tell.  The remove-me logic goes through getFilteredLists(), and that works on the key to peopleToLabelToFiles, which is still the person's name.  (My previous way of doing this _did_ break remove-me, but this way should not.)

Why did you think this would impact remove-me?

**Comment by csilvers** [comment]
> You can also just test things in the Gerald repo, I think, if you want to avoid the long chain of gerald → actions → webapp.

Sadly, I can't, because the gerald repo's github action is `Khan/actions@gerald-pr-v3`, so it doesn't see the new code until I go through all the rigamarole.

File: src/utils.js (hunk 3 above)
Comment (lillialexis) [reply src/utils.js:59]: > Why did you think this would impact remove-me?
I guess I thought it might impact the remove-me function because I didn't really know how to remove me function works… I just had a hunch that it might, and I wanted to make sure that it was considered and tested.

Sounds like the way that the functionality works is that it just re-creates the comment minus the removed person? So if the removed person was originally in there twice, and they wanted to be removed, it just leave them out for both rules? Instead of, say, matching the name in the body of the previous comment.

**Comment by lillialexis** [comment]
@csilvers I believe you can temporarily point Gerald to the sha that is one behind of the tip of your PR? Because the sha of the commit that points to one behind the tip is the tip...

File: src/utils.js (hunk 3 above)
Comment (csilvers) [reply src/utils.js:59]: That's right.

**Comment by csilvers** [comment]
> @csilvers I believe you can temporarily point Gerald to the sha that is one behind of the tip of your PR? Because the sha of the commit that points to one behind the tip is the tip...

Oh, I can change the label.  But that will change things for everyone, not just me, so it's a little scary.  I'll try it later in the day.

**Comment by lillialexis** [comment]
@csilvers I think on this branch, point this repo's workflow file to a sha on this branch. I think that should work in a way that doesn't affect other people?

**Comment by csilvers** [comment]
> @csilvers I think on this branch, point this repo's workflow file to a sha on this branch. I think that should work in a way that doesn't affect other people?

I don't know how to do that.  Right now the repo's gerald-pr.yml file looks like this:
```
    steps:
      - uses: Khan/actions@gerald-pr-v3
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          admin-token: ${{ secrets.KHAN_ACTIONS_BOT_TOKEN }}
```
what are you proposing I change it to?

**Comment by csilvers** [comment]
@MiguelCastillo you're just being added because I needed a name for testing this new functionality.

There's one more test you could help with -- could you add a comment to this PR that just says `#removeme`?  I want to test that this then auto-removes you as a reviewer.

**Comment by MiguelCastillo** [comment]
> @MiguelCastillo you're just being added because I needed a name for testing this new functionality.
> 
> There's one more test you could help with -- could you add a comment to this PR that just says `#removeme`? I want to test that this then auto-removes you as a reviewer.

Sure thang!

#removeme

File: src/utils.js (hunk 3 above)
Comment (MiguelCastillo) [reply src/utils.js:59]: #removeme

**Comment by MiguelCastillo** [comment]
Not sure how important it is, but the tricky thing about using `#` to start a trigger word is that github thinks you want to reference a PR.  Just an annoying observation :D

Screenshot 2024-11-13 at 4 38 30 PM

**Comment by csilvers** [comment]
@MiguelCastillo : I think you may need to say `#removeme` here, at the top level.

**Comment by MiguelCastillo** [comment]
> @MiguelCastillo : I think you may need to say `#removeme` here, at the top level.

I did above.  I also did it as an inline comment in case that made a difference.  Will do a standalone one too.

**Comment by MiguelCastillo** [comment]
#removeme

**Comment by csilvers** [comment]
Oh, it's still processing the PR, maybe it will be a while before it's ready to run the github actions to do the update.

**Comment by lillialexis** [comment]
> > @csilvers I think on this branch, point this repo's workflow file to a sha on this branch. I think that should work in a way that doesn't affect other people?
> 
> I don't know how to do that. Right now the repo's gerald-pr.yml file looks like this:
> 
> ```
>     steps:
>       - uses: Khan/actions@gerald-pr-v3
>         with:
>           token: ${{ secrets.GITHUB_TOKEN }}
>           admin-token: ${{ secrets.KHAN_ACTIONS_BOT_TOKEN }}
> ```
> 
> what are you proposing I change it to?

@csilvers I tried and failed :-( https://github.com/Khan/gerald/pull/91

**Comment by csilvers** [comment]
It's still processing the PR.  Something is very wrong.  Probably the huge index.js file.

**Comment by csilvers** [comment]
OK, I'm landing this.  I'm not convinced `#removeme` works, but I have no experience with it, so I can't say for sure, or if not if it's this PR that broke it or not.  If it's being problematic in a way that matters I'm sure we'll hear about it!
